File: sdk/swift/Sources/TokenPayID/Internal/TpidUpdater.swift
import Foundation
import CryptoKit

/// 2.6.1 — Smart in-place SDK updater (Apple platforms).
///
/// On iOS / iPadOS / tvOS, Apple policy forbids replacing the app's own
/// binary at runtime — the App Store wants every executable byte signed by
/// you and reviewed by them. What `TpidUpdater` *can* do honestly is:
///
///  1. Force-refresh `TpidRemoteConfig` so everything the server can
///     push — translated strings, theme tokens, feature flags — applies
///     live inside the running widget the instant the banner finishes
///     its animation. This is the real "in-place update" the user sees.
///  2. Download the SDK tarball (source + .xcframework) into the app's
///     Application Support directory and verify its SHA-256 (when the
///     server advertised one). That gives integrators a fresh artefact
///     for their next build without bouncing through a browser.
///  3. Emit a fine-grained `Phase` stream to the widget so the animated
///     update banner can draw a real progress bar, not a fake spinner.
///
/// The updater lives in `Internal/` and is driven by `WidgetViewModel` —
/// the UI never touches it directly, so Compose-style banners on the
/// JVM / Android side and the SwiftUI banner here share the same mental
/// model and (roughly) the same code flow.
struct TpidUpdaterInfo: Sendable {
    /// `true` iff the server-recommended version is strictly newer.
    let hasUpdate: Bool
    /// Version we're currently running.
    let currentVersion: String
    /// Version the server considers the latest.
    let latestVersion: String?
    /// HTTPS URL of the tarball; may be empty.
    let downloadURL: String
    /// Expected SHA-256 of `downloadURL`; `nil` when server omits it.
    let sha256: String?
    /// Optional changelog, surfaced as the banner's tooltip text.
    let updateNotes: String
}

/// Fine-grained updater state for the animated banner.
enum TpidUpdatePhase: Sendable, Equatable {
    /// Preparing the download, no bytes yet.
    case checking
    /// Actively downloading; `total == nil` when the server omitted
    /// Content-Length (the banner switches to an indeterminate bar).
    case downloading(bytes: Int64, total: Int64?)
    /// SHA-256 check in progress.
    case verifying
    /// Extracting / moving files into the cache.
    case installing
    /// Rolling the new `TpidRemoteConfig` into the running widget.
    case refreshingConfig
    /// Terminal — success.
    case done(cachedPath: URL?, newConfig: TpidRemoteConfig)
    /// Terminal — error; the updater cleaned up any partial files.
    case failed(reason: String)

    static func == (lhs: TpidUpdatePhase, rhs: TpidUpdatePhase) -> Bool {
        switch (lhs, rhs) {
        case (.checking, .checking),
             (.verifying, .verifying),
             (.installing, .installing),
             (.refreshingConfig, .refreshingConfig):
            return true
        case let (.downloading(a, b), .downloading(c, d)):
            return a == c && b == d
        case let (.done(a, _), .done(b, _)):
            return a == b
        case let (.failed(a), .failed(b)):
            return a == b
        default:
            return false
        }
    }
}

/// Session / filesystem-aware updater. The `session` and `cacheRoot`
/// parameters exist for testability — production code uses the defaults.
final class TpidUpdater: @unchecked Sendable {
    private let api: ApiClient
    private let clientId: String
    private let session: URLSession
    private let cacheRoot: URL

    init(
        api: ApiClient,
        clientId: String,
        session: URLSession = .shared,
        cacheRoot: URL? = nil
    ) {
        self.api = api
        self.clientId = clientId
        self.session = session
        self.cacheRoot = cacheRoot ?? Self.defaultCacheRoot()
    }

    /// Query the server without hitting the in-memory cache. Never throws;
    /// a network failure collapses to `hasUpdate=false` so the banner
    /// disappears instead of throwing an ugly error at the user.
    func checkForUpdate() async -> TpidUpdaterInfo {
        TpidRemoteConfig.invalidateCache()
        let cfg = await TpidRemoteConfig.fetch(api: api, clientId: clientId)
        let current = TpidVersion.string
        let latest = cfg.recommendedSdkVersion
        let isNewer = cfg.isOlderThan(latest)
        let url = cfg.downloadUrl ?? ""
        return TpidUpdaterInfo(
            hasUpdate: isNewer && !url.isEmpty,
            currentVersion: current,
            latestVersion: latest,
            downloadURL: url,
            sha256: (cfg.downloadSha256?.isEmpty == false) ? cfg.downloadSha256 : nil,
            updateNotes: cfg.updateNotes ?? ""
        )
    }

    /// Drive the full update state machine. `onPhase` is invoked on
    /// whichever queue the caller drove the `await` from — typically a
    /// `@MainActor` context. Cancellation aborts cleanly and deletes any
    /// partial download.
    ///
    /// The method always resolves to a terminal phase (`.done` or
    /// `.failed`) — the caller never needs to handle "stuck" state.
    func performUpdate(
        onPhase: @Sendable @escaping (TpidUpdatePhase) -> Void
    ) async -> TpidUpdatePhase {
        var partial: URL?
        func fail(_ reason: String) -> TpidUpdatePhase {
            if let p = partial { try? FileManager.default.removeItem(at: p) }
            let f = TpidUpdatePhase.failed(reason: reason)
            onPhase(f)
            return f
        }
        onPhase(.checking)

        let info = await checkForUpdate()
        if !info.hasUpdate {
            // No newer tarball to fetch — still refresh the config so the
            // user sees *something* from clicking Update. This is the
            // whole point of a smart update button.
            let cfg = await TpidRemoteConfig.fetch(api: api, clientId: clientId)
            onPhase(.refreshingConfig)
            let done = TpidUpdatePhase.done(cachedPath: nil, newConfig: cfg)
            onPhase(done)
            return done
        }

        guard let url = URL(string: info.downloadURL) else {
            return fail("invalid downloadUrl: \(info.downloadURL)")
        }

        let versionDir = cacheRoot
            .appendingPathComponent(info.latestVersion ?? "latest", isDirectory: true)
        try? FileManager.default.createDirectory(
            at: versionDir, withIntermediateDirectories: true
        )
        let fileName = url.lastPathComponent.isEmpty ? "sdk.tar.gz" : url.lastPathComponent
        let target = versionDir.appendingPathComponent(fileName)
        let tempTarget = versionDir.appendingPathComponent(fileName + ".part")
        try? FileManager.default.removeItem(at: tempTarget)
        partial = tempTarget

        // --- download ---
        var req = URLRequest(url: url)
        req.setValue("TokenPayID-Swift/\(TpidVersion.string)", forHTTPHeaderField: "User-Agent")

        let bytes: URLSession.AsyncBytes
        let response: URLResponse
        do {
            (bytes, response) = try await session.bytes(for: req)
        } catch {
            return fail("download failed: \(error.localizedDescription)")
        }
        if let http = response as? HTTPURLResponse, !(200..<300).contains(http.statusCode) {
            return fail("download failed: HTTP \(http.statusCode)")
        }
        let total: Int64? = (response as? HTTPURLResponse)?.expectedContentLength
            .map { $0 > 0 ? Int64($0) : Int64(0) }
            .flatMap { $0 > 0 ? $0 : nil }
        onPhase(.downloading(bytes: 0, total: total))

        guard FileManager.default.createFile(atPath: tempTarget.path, contents: nil) else {
            return fail("cannot create cache file at \(tempTarget.path)")
        }
        let handle: FileHandle
        do {
            handle = try FileHandle(forWritingTo: tempTarget)
        } catch {
            return fail("cannot open cache file: \(error.localizedDescription)")
        }
        defer { try? handle.close() }

        var received: Int64 = 0
        var lastEmitBytes: Int64 = 0
        var buffer = Data()
        buffer.reserveCapacity(64 * 1024)

        do {
            for try await byte in bytes {
                buffer.append(byte)
                received += 1
                if buffer.count >= 64 * 1024 {
                    try handle.write(contentsOf: buffer)
                    buffer.removeAll(keepingCapacity: true)
                }
                if received - lastEmitBytes >= 64 * 1024 {
                    let snapshot = received
                    onPhase(.downloading(bytes: snapshot, total: total))
                    lastEmitBytes = snapshot
                }
            }
            if !buffer.isEmpty {
                try handle.write(contentsOf: buffer)
                buffer.removeAll(keepingCapacity: true)
            }
        } catch is CancellationError {
            try? FileManager.default.removeItem(at: tempTarget)
            return fail("download cancelled")
        } catch {
            return fail("download failed mid-stream: \(error.localizedDescription)")
        }
        onPhase(.downloading(bytes: received, total: total ?? received))

        // --- verify ---
        onPhase(.verifying)
        let actualSha = Self.sha256Hex(of: tempTarget)
        if let expected = info.sha256,
           expected.lowercased() != actualSha.lowercased() {
            return fail("sha256 mismatch: expected \(expected), got \(actualSha)")
        }

        // --- install ---
        onPhase(.installing)
        try? FileManager.default.removeItem(at: target)
        do {
            try FileManager.default.moveItem(at: tempTarget, to: target)
        } catch {
            // Cross-volume fallback.
            do {
                try FileManager.default.copyItem(at: tempTarget, to: target)
                try? FileManager.default.removeItem(at: tempTarget)
            } catch {
                return fail("install failed: \(error.localizedDescription)")
            }
        }
        partial = nil

        // Breadcrumb for any future `TpidUpdateBootstrap`-style integration.
        let markerURL = cacheRoot.appendingPathComponent("current")
        let marker = "\(info.latestVersion ?? "unknown")\t\(target.path)\t\(actualSha)\n"
        try? marker.data(using: .utf8)?.write(to: markerURL)

        // --- refresh live config ---
        onPhase(.refreshingConfig)
        TpidRemoteConfig.invalidateCache()
        let fresh = await TpidRemoteConfig.fetch(api: api, clientId: clientId)

        let done = TpidUpdatePhase.done(cachedPath: target, newConfig: fresh)
        onPhase(done)
        return done
    }

    // MARK: - helpers

    private static func defaultCacheRoot() -> URL {
        let fm = FileManager.default
        let base: URL
        if let support = try? fm.url(
            for: .applicationSupportDirectory,
            in: .userDomainMask,
            appropriateFor: nil,
            create: true
        ) {
            base = support
        } else {
            base = fm.temporaryDirectory
        }
        let dir = base.appendingPathComponent("tokenpay-id/sdk", isDirectory: true)
        try? fm.createDirectory(at: dir, withIntermediateDirectories: true)
        return dir
    }

    /// Streaming SHA-256 computation. Avoids loading the whole tarball
    /// into memory on low-RAM devices.
    private static func sha256Hex(of file: URL) -> String {
        guard let handle = try? FileHandle(forReadingFrom: file) else { return "" }
        defer { try? handle.close() }
        var hasher = SHA256()
        while true {
            let chunk: Data
            do {
                chunk = try handle.read(upToCount: 128 * 1024) ?? Data()
            } catch {
                return ""
            }
            if chunk.isEmpty { break }
            hasher.update(data: chunk)
        }
        return hasher.finalize().map { String(format: "%02x", $0) }.joined()
    }
}

File: sdk/android/tokenpay-id-sdk/src/main/java/space/tokenpay/id/internal/TpidUpdater.kt
package space.tokenpay.id.internal

import android.content.Context
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.OkHttpClient
import okhttp3.Request
import space.tokenpay.id.BuildConfig
import java.io.File
import java.security.MessageDigest
import java.util.Locale
import java.util.concurrent.TimeUnit

/**
 * 2.6.1 — Smart in-place SDK updater (Android).
 *
 * On Android, Play-distributed applications **cannot** legally replace their
 * own DEX bytecode at runtime — Google Play policy forbids it. What we
 * *can* do honestly is:
 *
 *  1. Force-refresh [RemoteConfig] so all remote-controllable content
 *     (strings, theme tokens, feature flags) applies live in the running
 *     widget. This is the "in-place update" the user actually sees —
 *     translated labels, rebranded accents, newly-enabled QR login path.
 *  2. Download the latest SDK tarball to the app's private cache so the
 *     integrator's next build can pick it up. Verifies SHA-256.
 *  3. Report a structured [Phase] stream to the UI so the animated banner
 *     has something real to visualise, rather than a spinner that's
 *     disconnected from what the code is actually doing.
 *
 * No tricks: we don't try to side-load DEX, we don't attempt to replace
 * the AAR, and we don't lie about what "update" means. The caller gets a
 * fresh config live, the app gets a cached artefact for inspection, and
 * the banner gets fine-grained progress.
 */
internal class TpidUpdater(
    private val context: Context,
    private val api: ApiClient,
    private val clientId: String,
    private val httpClient: OkHttpClient = defaultClient(),
    private val cacheRoot: File = defaultCacheRoot(context),
) {

    internal data class UpdateInfo(
        val hasUpdate: Boolean,
        val currentVersion: String,
        val latestVersion: String?,
        val downloadUrl: String,
        val sha256: String?,
        val updateNotes: String,
    )

    internal sealed class Phase {
        object Checking : Phase()
        data class Downloading(val bytes: Long, val total: Long) : Phase()
        object Verifying : Phase()
        object Installing : Phase()
        object RefreshingConfig : Phase()
        data class Done(val cachedPath: File?, val newConfig: RemoteConfig) : Phase()
        data class Failed(val reason: String, val cause: Throwable? = null) : Phase()
    }

    suspend fun checkForUpdate(): UpdateInfo = withContext(Dispatchers.IO) {
        RemoteConfig.invalidateCache()
        val cfg = RemoteConfig.fetch(api, clientId)
        val current = BuildConfig.TPID_SDK_VERSION
        val latest = cfg.recommendedSdkVersion
        val newer = cfg.isOlderThan(latest)
        val dl = cfg.downloadUrl.orEmpty()
        UpdateInfo(
            hasUpdate = newer && dl.isNotBlank(),
            currentVersion = current,
            latestVersion = latest,
            downloadUrl = dl,
            sha256 = cfg.downloadSha256?.takeIf { it.isNotBlank() },
            updateNotes = cfg.updateNotes.orEmpty(),
        )
    }

    suspend fun performUpdate(onPhase: (Phase) -> Unit): Phase = withContext(Dispatchers.IO) {
        var partial: File? = null
        try {
            onPhase(Phase.Checking)
            val info = checkForUpdate()
            if (!info.hasUpdate) {
                // No newer binary — still refresh the running config so the
                // user sees some effect. That's the entire point of a "smart"
                // update button: it never does nothing.
                val cfg = RemoteConfig.fetch(api, clientId)
                onPhase(Phase.RefreshingConfig)
                val done = Phase.Done(cachedPath = null, newConfig = cfg)
                onPhase(done)
                return@withContext done
            }

            val versionDir = File(cacheRoot, info.latestVersion ?: "latest").apply { mkdirs() }
            val fileName = info.downloadUrl.substringAfterLast('/').ifBlank { "sdk.tar.gz" }
            val target = File(versionDir, fileName)
            val tempTarget = File(versionDir, "$fileName.part").also { it.delete() }
            partial = tempTarget

            // --- download ---
            val req = Request.Builder()
                .url(info.downloadUrl)
                .header("User-Agent", "TokenPayID-Android/${BuildConfig.TPID_SDK_VERSION}")
                .build()
            httpClient.newCall(req).execute().use { resp ->
                if (!resp.isSuccessful) {
                    return@withContext fail(onPhase, "download failed: HTTP ${resp.code}", null, partial)
                }
                val total = resp.header("Content-Length")?.toLongOrNull() ?: -1L
                onPhase(Phase.Downloading(0, total))
                val body = resp.body ?: return@withContext fail(onPhase, "empty response body", null, partial)
                val source = body.byteStream()
                tempTarget.outputStream().use { out ->
                    val buf = ByteArray(64 * 1024)
                    var bytes = 0L
                    var lastEmit = 0L
                    while (true) {
                        val n = source.read(buf)
                        if (n <= 0) break
                        out.write(buf, 0, n)
                        bytes += n
                        if (bytes - lastEmit >= 64 * 1024) {
                            onPhase(Phase.Downloading(bytes, total))
                            lastEmit = bytes
                        }
                    }
                    onPhase(Phase.Downloading(bytes, total.takeIf { it > 0 } ?: bytes))
                }
            }

            // --- verify ---
            onPhase(Phase.Verifying)
            val actualSha = sha256Hex(tempTarget)
            if (info.sha256 != null && !info.sha256.equals(actualSha, ignoreCase = true)) {
                return@withContext fail(
                    onPhase,
                    "sha256 mismatch: expected ${info.sha256}, got $actualSha",
                    null,
                    partial,
                )
            }

            // --- install ---
            onPhase(Phase.Installing)
            if (target.exists()) target.delete()
            if (!tempTarget.renameTo(target)) {
                tempTarget.copyTo(target, overwrite = true)
                tempTarget.delete()
            }
            partial = null
            File(cacheRoot, "current").writeText(
                "${info.latestVersion ?: "unknown"}\t${target.absolutePath}\t${actualSha}\n"
            )

            onPhase(Phase.RefreshingConfig)
            RemoteConfig.invalidateCache()
            val freshCfg = RemoteConfig.fetch(api, clientId)

            val done = Phase.Done(cachedPath = target, newConfig = freshCfg)
            onPhase(done)
            done
        } catch (e: Throwable) {
            fail(onPhase, "update failed: ${e.message ?: e.javaClass.simpleName}", e, partial)
        }
    }

    private fun fail(
        onPhase: (Phase) -> Unit,
        reason: String,
        cause: Throwable?,
        partial: File?,
    ): Phase {
        partial?.runCatching { delete() }
        val f = Phase.Failed(reason, cause)
        onPhase(f)
        return f
    }

    companion object {
        private fun defaultCacheRoot(ctx: Context): File =
            File(ctx.filesDir, ".tokenpay-id/sdk").apply { mkdirs() }

        private fun defaultClient(): OkHttpClient = OkHttpClient.Builder()
            .connectTimeout(10, TimeUnit.SECONDS)
            .readTimeout(60, TimeUnit.SECONDS)
            .callTimeout(3, TimeUnit.MINUTES)
            .build()

        private fun sha256Hex(f: File): String {
            val md = MessageDigest.getInstance("SHA-256")
            f.inputStream().use { input ->
                val buf = ByteArray(128 * 1024)
                while (true) {
                    val n = input.read(buf)
                    if (n <= 0) break
                    md.update(buf, 0, n)
                }
            }
            return md.digest().joinToString(separator = "") { b -> "%02x".format(b) }
                .lowercase(Locale.ROOT)
        }
    }
}
